R/manage.R

```diff
@@ -0,0 +1,804 @@
+      rlang::abort(
+        "`version` must be an character version number or an integer."
+      )
+    }
+    version <- as.character(version)
+  }
+
+  if (length(available_versions) == 0) {
+    return(NULL)
+  }
+
+  if (is.null(available_versions)) {
+    available_versions <- unique(chrome_get_versions()$version)
+  }
+
+  if (identical(version, "latest")) {
+    return(max(available_versions))
+  }
+
+  available_versions <- numeric_version(unique(available_versions))
+
+  version_parts <- unclass(numeric_version(version))[[1]]
+
+  max_version <- rep(99999, 4)
+  max_version[seq_along(version_parts)] <- version_parts
+  max_version <- numeric_version(paste(max_version, collapse = "."))
+
+  min_version <- rep(0, 4)
+  min_version[seq_along(version_parts)] <- version_parts
+  min_version <- numeric_version(paste(min_version, collapse = "."))
+
+  available_versions <- available_versions[
+    available_versions <= max_version &
+      available_versions >= min_version
+  ]
+
+  if (length(available_versions) == 0) {
+    return(NULL)
+  }
+
+  max(available_versions)
+}
+
+curl_fetch_headers <- function(url) {
+  h <- curl::new_handle()
+  curl::handle_setopt(h, nobody = TRUE)
+  req <- curl::curl_fetch_memory(url, handle = h)
+  req_parse_headers(req)
+}
+
+req_parse_headers <- function(req) {
+  headers <- rawToChar(req$headers)
+  parsed_headers <- strsplit(headers, "\r\n")[[1]]
+  parsed_headers <- parsed_headers[parsed_headers != ""]
+  parsed_headers <- strsplit(parsed_headers, ": ")
+  parsed_headers <- rlang::set_names(
+    lapply(parsed_headers, `[`, 2),
+    sapply(parsed_headers, `[`, 1)
+  )
+
+  parsed_headers
+}
+
+req_headers_last_modified <- function(headers) {
+  names(headers) <- tolower(names(headers))
+  withr::with_locale(new = c("LC_TIME" = "en_US"), {
+    last_modified <- as.POSIXct(
+      headers[["last-modified"]],
+      format = "%a, %d %b %Y %H:%M:%S GMT",
+      tz = "GMT"
+    )
+    last_modified
+  })
+}
+
+download_json_cached <- function(url, update_cached = TRUE) {
+  path_cache <- chromote_cache_path()
+  dir.create(path_cache, showWarnings = FALSE, recursive = TRUE)
+
+  path_local <- file.path(path_cache, basename(url))
+
+  # Check if local file exists and get its modified time
+  if (file.exists(path_local)) {
+    if (!update_cached) {
+      return(path_local)
+    }
+
+    local_mtime <- file.info(path_local)$mtime
+
+    is_local_stale <- tryCatch(
+      {
+        # Fetch headers from the server
+        headers <- curl_fetch_headers(url)
+        server_last_modified <- req_headers_last_modified(headers)
+
+        length(server_last_modified) == 1 && local_mtime < server_last_modified
+      },
+      error = function(err) {
+        rlang::inform(
+          "Could not reach Chrome for Testing to update available versions.",
+          parent = err
+        )
+        FALSE
+      }
+    )
+
+    # Compare local file time with server's last-modified
+    if (is_local_stale) {
```

Comment:
Should this be _if local is not stale..._?
```suggestion
    if (!is_local_stale) {
```

Reply:
Yeah. I think I went back and forth about the naming and missed this. Thanks!